### PR TITLE
HTS-1210 Implementated mongo to spin up within test code

### PR DESCRIPTION
**Pull request by Jackie1988**

### CODE REVIEW

File: conf/application.conf [hunk 1]
```diff
@@ -234,7 +234,7 @@ eligibility-stats {
 # let min-backoff be `m` and max-backoff be `M`. Then `2m-M` cannot equal zero
 uc-threshold {
   # this is a temporary flag to disable the DES theshold behaviour until we are ready in production
-  enabled     = false
+  enabled     = true
```
Comment (AndyHWChung) [inline conf/application.conf]: is there a reason why this has changed in this PR?

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala [hunk 2]
```diff
@@ -40,111 +43,123 @@ class MongoEmailStoreSpec extends TestSupport with MongoTestSupport[EmailData, M
...
-    }
-
-    override def doUpdate(encryptedEmail: String, nino: NINO)(implicit ec: ExecutionContext): Future[Option[EmailData]] =
-      mockDBFunctions.update(EmailData(nino, encryptedEmail))
-
-    override def find(query: (String, Json.JsValueWrapper)*)(implicit ec: ExecutionContext): Future[List[EmailData]] =
-      query.toList match {
-        case head :: Nil ⇒ mockDBFunctions.get(head._2)
-        case _           ⇒ fail("Find function expected only 1 parameter")
+      override def indexes: Seq[Index] = {
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: do we still need this?

File: conf/application.conf (hunk 1 above)
Comment (Jackie1988) [reply conf/application.conf]: no I don't know how that happened, I've changed it back

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala (hunk 2 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: No we don't :)

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala [hunk 3]
```diff
@@ -16,46 +16,42 @@
...
+      val document = Json.obj("eligibilityReason" -> 7, "source" -> "Digital", "total" -> 1).value
 
       "return results as expected" in {
-        mockAggregate(Right(AggregationFramework.AggregationResult(documents)))
-        Await.result(mongoStore.getEligibilityStats(), 5.seconds) shouldBe List(EligibilityStats(Some(7), Some("Digital"), 1))
+        withMongo { reactiveMongoComponent ⇒
+          val store = newEligibilityStatsMongoStore(reactiveMongoComponent)
+
+          store.collection.insert(document)
+          Await.result(store.doAggregate(), 5.seconds)
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: is this line required ?

File: build.sbt [hunk 4]
```diff
@@ -147,3 +148,6 @@ lazy val microservice = Project(appName, file("."))
     Resolver.jcenterRepo,
     Resolver.bintrayRepo("hmrclt", "maven")
   ))
+  .settings(
+    testOptions in Test += Tests.Argument("-oDF")
```
Comment (sureshhmrc) [inline build.sbt]: why is this required ?

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala [hunk 5]
```diff
@@ -16,46 +16,42 @@
...
+        }
       }
 
       "handle error while reading from mongo" in {
-        mockAggregate(Left("unexpected error"))
-        Await.result(mongoStore.getEligibilityStats(), 5.seconds) shouldBe List.empty
+        withMongo { reactiveMongoComponent ⇒
+          val store = newEligibilityStatsMongoStore(reactiveMongoComponent)
+
+          store.doAggregate()
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: is this line required ?

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala [hunk 6]
```diff
@@ -40,111 +43,115 @@ class MongoEmailStoreSpec extends TestSupport with MongoTestSupport[EmailData, M
...
+            mockEncrypt(email)(encryptedEmail)
+            mockDecrypt(encryptedEmail)(Some(email))
+          }
+
+          update(nino, email, emailStore)
+
+          val storedEmail = getConfirmedEmail(nino, emailStore)
+          storedEmail shouldBe Right(Some(email))
+
+          emailStore.collection.db.connection.askClose()
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: do we need this here ? if so we its not there in other tests ?

File: test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala [hunk 7]
```diff
@@ -50,23 +38,19 @@ class UserCapStoreSpec extends TestSupport with MongoTestSupport[UserCap, MongoU
     "getting the user-cap" should {
 
       "return the existing record successfully" in {
-
-        mockGet()(toFuture(Some(record)))
-        Await.result(mongoStore.get(), 5.seconds) shouldBe Some(record)
+        withMongo { reactiveMongoComponent ⇒
+          val store = newUserCapMongoStore(reactiveMongoComponent)
+          Await.result(store.doUpdate(record), 5.seconds)
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala]: you are not asserting the return value here ?

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala [hunk 8]
```diff
@@ -0,0 +1,95 @@
+/*
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: you can move this to `repo` package

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    eventually{
+      mongodProps.mongodProcess.isProcessRunning shouldBe false
+    }
+  }
+
+  private def mongo(mongodProps: MongodProps, failoverStrategy: Option[FailoverStrategy] = None): ReactiveMongoComponent = new ReactiveMongoComponent {
+    override def mongoConnector: MongoConnector = {
+      val address = mongodProps.mongodProcess.getConfig.net()
+      MongoConnector(
+        s"mongodb://${address.getServerAddress.getHostAddress}:${address.getPort}/test",
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: why is `test` db here , but not `help-to-save` ?

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    val port = {
+      val socket = new ServerSocket(0)
+      socket.close()
+
+      eventually {
+        socket.isClosed shouldBe true
+      }
+
+      socket.getLocalPort
+    }
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: can you clarify why the above port logic is needed ?  you are reading `getLocalPort` from a closed socket, so not how it works

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala (hunk 3 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: No as getEligibilityStats already calls this method, so I have removed this line

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala (hunk 5 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: removed

File: build.sbt (hunk 4 above)
Comment (Jackie1988) [reply build.sbt]: we were trying to obtain more of the stack trace but Andy and I don't think this made any difference, so removing ...

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala (hunk 6 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: I've removed and the tests still passed, so its not needed

File: test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala (hunk 7 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala]: I've added an assert on the return value

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 8 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: I think utils is more suited?

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 9 above)
Comment (sureshhmrc) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: yeah, but name can be something meaningful,

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 8 above)
Comment (sureshhmrc) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: its being used by only mongo specs, so no point in keeping in util package

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 10 above)
Comment (sureshhmrc) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: is there any chance that some other process grabs this port before you pick it ?

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala (hunk 5 above)
Comment (sureshhmrc) [reply test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: i don't see the commit after PR feedback ?  have you pushed it already ?

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 10 above)
Comment (AndyHWChung) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: it's a small chance - something will have to grab the port in the milliseconds in between when we give up the port and mongo picking it up

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala (hunk 5 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: no not yet

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 9 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: ok I've renamed it to help-to-save

File: test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala (hunk 8 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/utils/MongoSupport.scala]: ok I've moved it.

File: test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala [hunk 11]
```diff
@@ -16,46 +16,41 @@
...
 import uk.gov.hmrc.helptosave.repo.MongoEligibilityStatsStore.EligibilityStats
-import uk.gov.hmrc.helptosave.repo.MongoEnrolmentStore.EnrolmentData
 import uk.gov.hmrc.helptosave.utils.TestSupport
 
+import scala.concurrent.Await
 import scala.concurrent.duration._
-import scala.concurrent.{Await, Future}
 
-class EligibilityStatsStoreSpec extends TestSupport with MongoTestSupport[EnrolmentData, MongoEligibilityStatsStore] {
+class EligibilityStatsStoreSpec extends TestSupport with Eventually with MongoSupport {
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/EligibilityStatsStoreSpec.scala]: are you using Eventually here ?  if not remove it and import also

File: test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala [hunk 12]
```diff
@@ -50,23 +38,19 @@ class UserCapStoreSpec extends TestSupport with MongoTestSupport[UserCap, MongoU
...
-    "updating the user-cap" should {
-
-      "return the updated record successfully" in {
-
-        mockUpdate(record)(Right(Some(record)))
-        Await.result(mongoStore.upsert(record), 5.seconds) shouldBe Some(record)
+        withMongo { reactiveMongoComponent ⇒
+          val store = newUserCapMongoStore(reactiveMongoComponent)
+          store.removeAll()
+          Await.result(store.get(), 5.seconds) shouldBe None
```
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala]: check if you can define a method in TestSupport something like 
``` def await[T](f: Future[T]): T = {
        Await.result(f, 5.seconds)
      } `` so that you can use it every where

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala [hunk 13]
```diff
@@ -40,111 +43,113 @@ class MongoEmailStoreSpec extends TestSupport with MongoTestSupport[EmailData, M
...
     "updating emails" must {
 
-      val nino = "NINO"
-      val email = "EMAIL"
-      val encryptedEmail = "ENCRYPTED"
       val data = EmailData(nino, encryptedEmail)
 
-        def update(nino: NINO, email: String): Either[String, Unit] =
-          Await.result(mongoStore.storeConfirmedEmail(email, nino).value, 5.seconds)
+        def update(nino: NINO, email: String, emailStore: MongoEmailStore): Either[String, Unit] =
```
[the remark this answers is not in the document]
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: I've renamed this, thanks

File: test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala [hunk 14]
```diff
@@ -40,111 +43,113 @@ class MongoEmailStoreSpec extends TestSupport with MongoTestSupport[EmailData, M
...
-        // now try when there is no email
-        mockFind(nino)(Future.successful(List()))
-        get(nino) shouldBe Right(None)
+          val emailStore = newMongoEmailStore(reactiveMongoComponent)
+
+          //there should no emails to start with
+          get(nino, emailStore) shouldBe Right(None)
+
+          //putting the email in mongo
+          Await.result(emailStore.storeConfirmedEmail(email, nino).value, 5.seconds)
```
[the remark this answers is not in the document]
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/MongoEmailStoreSpec.scala]: yeah ok

File: test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala (hunk 12 above)
Comment (Jackie1988) [reply test/uk/gov/hmrc/helptosave/repo/UserCapStoreSpec.scala]: I tried but the tests resulted in futures not complete
